### ISSUE TITLE
Optimize string allocation in the Tokenizer

### ISSUE THREAD

**Issue by Yves57** [aspnet/Razor#873]
I have seen that many strings are allocated in the `Microsoft.AspNetCore.Razor.Tokenizer.Tokenizer.EndSymbol()` method.
But some strings are in fact often the same. So I modified the end of the method like that:

```
.....
string tokenContent;
if (Buffer.Length != 1)
{
   tokenContent = Buffer.ToString();
}
else
{
   if (!_oneCharTokenCache.TryGetValue(Buffer[0], out tokenContent))
   {
      tokenContent = Buffer.ToString();
      _oneCharTokenCache.Add(Buffer[0], tokenContent);
   }
}
sym = CreateSymbol(start, tokenContent, type, errors);
```

I have made a test with an extrem case ([LargePageMvc page](https://github.com/aspnet/Performance/blob/dev/testapp/LargePageMvc/Views/Home/Index.cshtml)).

Before the modification, 215643 strings where allocated. After the optimization, 131577 strings allocations where saved (more than half). And the `_oneCharTokenCache` dictionary had only 17 items inside: some symbols like <, !, >, =, and all digits.

Two additional remarks:
- As a bonus, it improves performances from 587ms to call `MarkupParser.ParseDocument()` down to 510ms on my (old!) computer. It can improve a little bit F5 startup scenarios with large pages...
- Of course, we can extend the concept to any `string` lengths, but there is less redundancy, and it will be probably less performant.

If you think that it is interesting, I can propose a PR...

**Comment by Yves57** [comment]
Other optimization to propose (but smaller).
In [CSharpTokenizer.Identifier](https://github.com/aspnet/Razor/blob/dev/src/Microsoft.AspNetCore.Razor/Tokenizer/Internal/CSharpTokenizer.cs#L477), the buffer is transformed to string two times.

**Comment by Yves57** [comment]
Still around `Tokenizer.EndSymbol()` and still to avoid redundant string allocations, I propose to add a `protected virtual string GetSymbolContent(TSymbolType type)` that manages the suggested dictionary proposed in my first issue comment.
But it would be overriden in `HtmlTokenizer` like that:

```
protected override string GetSymbolContent(HtmlSymbolType type)
{
    if (type == HtmlSymbolType.NewLine && Buffer.Length == 1)
    {
       return "\r";
    }
    else if (type == HtmlSymbolType.NewLine && Buffer.Length == 2)
    {
       return "\r\n";
    }
    else if (type == HtmlSymbolType.WhiteSpace && Buffer.Length <= 32)
    {
        return _whitespaceConstants[Buffer.Length - 1];
    }
    return super.GetSymbolContent(type);
}
```

In the same "LargePageMvc" test, it would save about 24213 more allocations (about half for spaces, and half for new lines).

PS: I hope my calculations are not wrong ;-).

**Comment by rynowak** [comment]
@Yves57 - this is a really great experiment, and something we're definitely interested in. This is something that the team has discussed doing in the past and it really doesn't have any downside considering that HTML is pretty repetitive and we're only considering the punctuation and tag/attribute names. If we did something like this we'd definitely take it further than 1-char strings.


In particular seeing the impact of this for 1-char string is surprising and seems like a good opportunity.

However, I'm not sure the best to proceed with this specific change because we're in the middle of remodeling right now (https://github.com/aspnet/Razor/issues/836).

Sit tight for a bit, I want to talk this over with a few people here and see if we have the potential to include this in a future release or if we should just plan on baking this into the new codebase.

**Comment by Yves57** [comment]
@rynowak - OK, I understand and wait your decision.

For the 1-char idea, I feel that the gain of 1-char dictionary is always visible and more predictable (small dictionary, each entry used many times). In a multi-char string case, the gain is less predictable: the dictionary may be huge, entries may be used a few times, etc. Sorry for the poor English, I'm not sure to be clear in my explanation :-).

Last thing for today: I think that many allocations can be suppressed in `LineTrackingStringBuffer` too (I think that it is possible to remove the `Append` method, remove the `StringBuilder`for each line, and keep only the input string). Again tens of thousands allocations avoided if everything works as I expect.

**Comment by rynowak** [comment]
Sorry for the delay, things slow down here around the holidays ⛄️ 

We're not going to be able to include this in a 1.1.1 release of the existing codebase, but we'd like to see it fixed for `Microsoft.AspNetCore.Razor.Evolution`. Currently the parser/tokenizer code in that library is pretty much verbatim from the existing shipped codebase (with the exception of being internal).

Feel free to send a PR, or series of PRs if you find more improvements. We're interested in any or all of your findings in this area.

**Comment by Eilon** [comment]
@rynowak I'm assigning to you until we figure out exactly how this fits in.

**Comment by Yves57** [comment]
Short term status:
- First PR #908 (checks fails, but I think that it is a general issue).
- I will look to optimize `LineTrackingStringBuffer` (I hope that I can remove the `StringBuilder` instances).

For future optimization, I was playing with the `TagHelperScopeManager / TagHelperExecutionContext` classes.
- I have seen in [issue MVC 3516](https://github.com/aspnet/Mvc/issues/3516#issuecomment-203647621) that using TagHelpers allocates many `Dictionary<object, object>` (I suppose for `TagHelperExecutionContext.Items`). I have not found any scenario that uses this `TagHelperExecutionContext.Items` property. In fact I don't know for which scenario it exists. So I was asking my self if the dictionary cound not be allocated only if necessary. I need your knowledge for this idea.
- `TagHelperExecutionContext._tagHelpers` is always allocated. I need your knowledge again (sorry :-), but what is the most frequent case: one TagHelper in this list (so we can think about a fast path), or several TagHelpers.

**Comment by NTaylorMullen** [comment]
- `TagHelperExecutionContext.Items` is used to communicate between TagHelper parent to child. Mvc uses it in a few scenarios, [here's](https://github.com/aspnet/Mvc/blob/eea297975db6e77a39b4809489a28ed65841584c/src/Microsoft.AspNetCore.Mvc.TagHelpers/SelectTagHelper.cs#L80) one of them.
- `TagHelperExecutionContext._tagHelpers` might be able to be optimized if we prioritize the 1 TagHelper scenario. It's what tracks the active TagHelpers that will run on a given HTML element.

**Comment by rynowak** [comment]
> I will look to optimize LineTrackingStringBuffer (I hope that I can remove the StringBuilder instances).

FYI this is on my hit-list to remove completely

**Comment by Yves57** [comment]
> FYI this is on my hit-list to remove completely

I suppose that it is related to your recent PR about 'line-mapping'. I remove it from my list 😄 .

**Comment by Yves57** [comment]
@NTaylorMullen 
Thank you for the explanation. Are you interested by optimizations (if I can find a solution!) for these two properties?

**Comment by rynowak** [comment]
Yeah, the plan here is to port the tokenizer code to be based around `RazorSourceDocument`. This is more similar to https://msdn.microsoft.com/en-us/library/microsoft.visualstudio.text.itextsnapshot.aspx which we work with in the editor, and the roslyn `SourceText` abstraction which we also want to interoperate with.

**Comment by rynowak** [comment]
Update a few months later, we're still not ready to dive into this, still stabilizing the backend and tooling
